### PR TITLE
[Bridge] Add State Transaction and state sync mechanism in IBFT

### PR DESCRIPTION
**Pull request by Kourin1996**
# Description

Please provide a detailed description of what was done in this PR

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: bridge/event.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+func ParseStateSyncEvent(log *web3.Log) (*StateSyncEvent, error) {
+	event, err := StateSyncedEvent.ParseLog(log)
+	if err != nil {
+		return nil, err
+	}
+
+	var id *big.Int
+	var contractAddr web3.Address
+	var data []byte
+	var ok bool
```
Comment (zivkovicmilos) [inline bridge/event.go]: Please group these under one `var (...)`

File: bridge/event.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+	contractAddr, ok = event["contractAddress"].(web3.Address)
+	if !ok {
+		return nil, errors.New("failed to parse contractAddress in StateSyncedEvent")
+	}
+
+	data, ok = event["data"].([]uint8)
+	if !ok {
+		return nil, errors.New("failed to parse data in StateSyncedEvent")
+	}
```
Comment (zivkovicmilos) [inline bridge/event.go:44]: What do you think about extracting these hardcoded values `id`, `data` and `contactAddress` into constants, and just referencing them whenever you need to?

File: bridge/message.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//}
+//
+//func (m *Message) MarshalRLPWith(arena *fastrlp.Arena) *fastrlp.Value {
+//	vv := arena.NewArray()
+//
+//	vv.Set(arena.NewBytes(m.ID.Bytes()))
+//	vv.Set(arena.NewBytes(m.Transaction.MarshalRLP()))
+//
+//	return vv
+//}
```
Comment (zivkovicmilos) [inline bridge/message.go]: Why is this commented out?

File: bridge/transaction.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+var (
+	ParsedStateReceiverABI = abi.MustNewABI(StateReceiverABI)
+)
+
+func NewStateSyncedTx(event *StateSyncEvent) *types.Transaction {
+	method := ParsedStateReceiverABI.Methods["onStateReceived"]
+	data, _ := method.Encode(map[string]interface{}{
+		"data": event.Data,
+	})
```
Comment (zivkovicmilos) [inline bridge/transaction.go:47]: Same as with https://github.com/0xPolygon/polygon-edge/pull/428/files#r817649103

File: consensus/ibft/ibft.go [hunk 5]
```diff
@@ -751,12 +785,8 @@ func (i *Ibft) runAcceptState() { // start new round
 	}
 
 	if i.bridge != nil {
+		// TODO: Check threshold
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Leftover todo

File: consensus/ibft/ibft.go [hunk 6]
```diff
@@ -866,6 +896,21 @@ func (i *Ibft) runAcceptState() { // start new round
...
+						continue
+					}
+
+					if err := i.bridge.ValidateTx(tx); err != nil {
+						i.logger.Error("block has the invalid state transaction", "tx", tx.Hash, "err", err)
+						i.handleStateErr(errBlockVerificationFailed)
+						break StateTxValidationLoop
+					}
+				}
+			}
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Please refactor this part of the code to not use labels, there is nothing elegant with writing code like it's the 1970s

File: consensus/ibft/ibft.go [hunk 7]
```diff
@@ -1281,6 +1328,19 @@ func (i *Ibft) GetBlockCreator(header *types.Header) (types.Address, error) {
...
+		for _, tx := range block.Transactions {
+			if tx.Type != types.TxTypeState {
+				continue
+			}
+
+			i.bridge.Consume(tx)
+		}
+	}
+}
+
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Leftover TODO

File: e2e/bridge_test.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+)
+
+func TestBridge_StateSync(t *testing.T) {
+	senderKey, senderAddr := tests.GenerateKeyAndAddr(t)
+
+	// Start root chain
+	sourceIBFT := framework.NewIBFTServersManager(
+		t,
+		IBFTMinNodes,
+		fmt.Sprintf("%ssource", IBFTDirPrefix),
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go:125]: Leftover print

File: e2e/bridge_test.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+	sourceIBFT.StartServers(ctx1)
+
+	// deploy state syncer contract
+	ctx2, cancel2 := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel2()
+	syncerContractAddr, err := sourceIBFT.GetServer(0).DeployContract(ctx2, StateSyncerByteCode, senderKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Printf("Deployed StateSyncerByteCode %s\n", syncerContractAddr)
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go]: Leftover print

File: e2e/bridge_test.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+	// setup destination chain
+	destIBFT := framework.NewIBFTServersManager(
+		t,
+		IBFTMinNodes,
+		fmt.Sprintf("%sdest", IBFTDirPrefix),
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go:149]: Leftover print

File: e2e/bridge_test.go [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+	destIBFT.StartServers(ctx3)
+
+	// deploy state reciever contract
+	ctx4, cancel4 := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel4()
+	reciverContractAddr, err := destIBFT.GetServer(0).DeployContract(ctx4, StateReceiverByteCode, senderKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+	fmt.Printf("Deployed StateSyncerContractAddr %s\n", reciverContractAddr)
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go]: Leftover print

File: e2e/bridge_test.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+	// Send state sync contract
+	ctx5, cancel5 := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel5()
+	to := types.BytesToAddress(syncerContractAddr[:])
+	input, err := StateSyncerABI.Methods["stateSync"].Encode(map[string]interface{}{
+		"contractAddress": reciverContractAddr.String(),
+		"data":            []byte{0x68, 0x65, 0x6c, 0x6c, 0x6f}, // hello
+	})
+	fmt.Printf("input %+v\n", input)
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go]: Leftover print

File: e2e/bridge_test.go [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    }
+  ]`
+)
+
+var (
+	StateSyncerABI   = abi.MustNewABI(StateSyncerABIJSON)
+	StateRecieverABI = abi.MustNewABI(StateReceiverABIJSON)
+)
+
+func TestBridge_StateSync(t *testing.T) {
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go:118]: Please avoid naming variables with `varName#` where `#` is a decimal number

File: e2e/bridge_test.go [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+		GasPrice: big.NewInt(framework.DefaultGasPrice),
+		To:       &to,
+		From:     senderAddr,
+		Input:    input,
+	}, senderKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO: remove of course
```
Comment (zivkovicmilos) [inline e2e/bridge_test.go]: This one made me laugh 🙂 

Remove, of course 💯 

cc @Vuksan

File: state/executor.go [hunk 15]
```diff
@@ -238,7 +238,8 @@ func (t *Transition) Write(txn *types.Transaction) error {
 	signer := crypto.NewSigner(t.config, uint64(t.r.config.ChainID))
 
 	var err error
-	if txn.From == emptyFrom {
+	// TODO: remove knowledge of TxTypeState
+	if txn.Type != types.TxTypeState && txn.From == emptyFrom {
```
Comment (zivkovicmilos) [inline state/executor.go]: Leftover TODO.

What does it mean?

File: state/executor.go [hunk 16]
```diff
@@ -256,7 +257,10 @@ func (t *Transition) Write(txn *types.Transaction) error {
 		return e
 	}
 
-	t.totalGas += result.GasUsed
+	// TODO: remove knowledge of TxTypeState
+	if txn.Type != types.TxTypeState {
+		t.totalGas += result.GasUsed
+	}
```
Comment (zivkovicmilos) [inline state/executor.go]: Same as https://github.com/0xPolygon/polygon-edge/pull/428/files#r817660320

File: state/executor.go [hunk 17]
```diff
@@ -471,25 +500,28 @@ func (t *Transition) apply(msg *types.Transaction) (*runtime.ExecutionResult, er
...
-	txn.AddBalance(t.ctx.Coinbase, coinbaseFee)
+		// pay the coinbase
+		coinbaseFee := new(big.Int).Mul(new(big.Int).SetUint64(result.GasUsed), gasPrice)
+		txn.AddBalance(t.ctx.Coinbase, coinbaseFee)
 
-	// return gas to the pool
-	t.addGasPool(result.GasLeft)
+		// return gas to the pool
+		t.addGasPool(result.GasLeft)
+	}
```
Comment (zivkovicmilos) [inline state/executor.go]: What do you think about extracting the `apply` method to be reimplemented by regular transactions and these TxTypeState transactions?

That way we can avoid this if spaghetti

File: txpool/txpool.go [hunk 18]
```diff
@@ -467,6 +468,11 @@ func (p *TxPool) processEvent(event *blockchain.Event) {
 // validateTx ensures the transaction conforms to specific
 // constraints before entering the pool.
 func (p *TxPool) validateTx(tx *types.Transaction) error {
+	// FIXME: Accepts only legacy transaction only for now
+	if tx.Type != types.TxTypeLegacy {
+		return ErrInvalidTxType
+	}
```
Comment (zivkovicmilos) [inline txpool/txpool.go:474]: I have no idea, leftover?

File: types/rlp_unmarshal.go [hunk 19]
```diff
@@ -303,7 +374,7 @@ func (t *Transaction) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
 		return fmt.Errorf("not enough elements to decode transaction, expected 9 but found %d", num)
 	}
 
-	p.Hash(t.Hash[:0], v)
+	// p.Hash(t.Hash[:0], v)
```
Comment (zivkovicmilos) [inline types/rlp_unmarshal.go]: Leftover?

File: types/transaction.go [hunk 20]
```diff
@@ -1,12 +1,45 @@
...
 )
 
+// TransactionType represents the value between 0x00 and 0x7F to represents the type of transaction
+// TransactionType is put before transaction data in RLP encoded
+type TransactionType uint8
+
+const (
+	TxTypeLegacy TransactionType = 0x0
+	// TxTypeAccessList TransactionType = 0x1
+	// TxTypeDynamicFee TransactionType = 0x2
```
Comment (zivkovicmilos) [inline types/transaction.go:18]: Why are they commented out?

File: types/transaction.go [hunk 21]
```diff
@@ -28,15 +63,24 @@ func (t *Transaction) IsContractCreation() bool {
...
-	v := t.MarshalRLPWith(ar)
-	hash.WriteRlp(t.Hash[:0], v)
+	//ar := marshalArenaPool.Get()
+	//hash := keccak.DefaultKeccakPool.Get()
+	//
+	//v := t.MarshalRLPWith(ar)
+	//hash.WriteRlp(t.Hash[:0], v)
+	//
+	//marshalArenaPool.Put(ar)
+	//keccak.DefaultKeccakPool.Put(hash)
```
Comment (zivkovicmilos) [inline types/transaction.go]: Leftover?

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
This PR is definitely not ready to be reviewed, there are a lot of things that need to get cleaned up before people should look at it

File: bridge/message.go (hunk 3 above)
Comment (Kourin1996) [reply bridge/message.go]: No need right now

File: consensus/ibft/ibft.go (hunk 5 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: This is not leftover. We need to decide what threshold is? 2/3 validators? or full validators?

File: e2e/bridge_test.go (hunk 8 above)
Comment (Kourin1996) [reply e2e/bridge_test.go:125]: This is sprintf

File: bridge/event.go (hunk 1 above)
Comment (Kourin1996) [reply bridge/event.go]: Grouped at 1933a26cdd08c99ef1f0544a99b6a1738231cc7d

File: bridge/message.go (hunk 3 above)
Comment (Kourin1996) [reply bridge/message.go]: Removed at 1933a26cdd08c99ef1f0544a99b6a1738231cc7d

File: bridge/event.go (hunk 2 above)
Comment (Kourin1996) [reply bridge/event.go:44]: We don't need to do this right now. But could in the future

File: e2e/bridge_test.go (hunk 9 above)
Comment (Kourin1996) [reply e2e/bridge_test.go]: Removed a0b920bafd73184156e9de23ba8634726836911e

File: e2e/bridge_test.go (hunk 10 above)
Comment (Kourin1996) [reply e2e/bridge_test.go:149]: This is Sprintf, You don't think if it's necessary?

File: e2e/bridge_test.go (hunk 11 above)
Comment (Kourin1996) [reply e2e/bridge_test.go]: Removed a0b920bafd73184156e9de23ba8634726836911e

File: e2e/bridge_test.go (hunk 12 above)
Comment (Kourin1996) [reply e2e/bridge_test.go]: Removed a0b920bafd73184156e9de23ba8634726836911e

File: e2e/bridge_test.go (hunk 13 above)
Comment (Kourin1996) [reply e2e/bridge_test.go:118]: Renamed a0b920bafd73184156e9de23ba8634726836911e

File: txpool/txpool.go (hunk 18 above)
Comment (Kourin1996) [reply txpool/txpool.go:474]: No, we need to fix when we add new type transaction besides StateTransaction

File: types/transaction.go (hunk 20 above)
Comment (Kourin1996) [reply types/transaction.go:18]: Now we don't support these transaction types but we may introduce in the future

File: types/transaction.go (hunk 21 above)
Comment (Kourin1996) [reply types/transaction.go]: Removed 957711a4531a42287ce327d8a4f0d5a7373572b5

File: types/rlp_unmarshal.go (hunk 19 above)
Comment (Kourin1996) [reply types/rlp_unmarshal.go]: Removed 957711a4531a42287ce327d8a4f0d5a7373572b5

File: bridge/event.go (hunk 1 above)
Comment (Kourin1996) [reply bridge/event.go]: Fixed at a0b920bafd73184156e9de23ba8634726836911e

File: consensus/ibft/ibft.go (hunk 7 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Moved 11e272b3658a425586dc95f8ba34bab2ec9cc563

File: consensus/ibft/ibft.go (hunk 6 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Removed when moving this flow to hook 11e272b3658a425586dc95f8ba34bab2ec9cc563

File: e2e/bridge_test.go (hunk 14 above)
Comment (Kourin1996) [reply e2e/bridge_test.go]: Fixed b908be88021f5c3c20443ee54ad1818a979f0825

File: state/executor.go (hunk 15 above)
Comment (Kourin1996) [reply state/executor.go]: Fixed eddbd35203739b6fe17c9e13481e676cf9fa36d4

File: state/executor.go (hunk 16 above)
Comment (Kourin1996) [reply state/executor.go]: Fixed eddbd35203739b6fe17c9e13481e676cf9fa36d4

File: state/executor.go (hunk 17 above)
Comment (Kourin1996) [reply state/executor.go]: Separated eddbd35203739b6fe17c9e13481e676cf9fa36d4

**Pull request merged**